lib/elixir/lib/kernel.ex

```diff
@@ -1599,19 +1599,63 @@ defmodule Kernel do
   # Extracts concatenations in order to optimize many
   # concatenations into one single clause.
   defp extract_concatenations({:<>, _, [left, right]}) do
-    [wrap_concatenation(left) | extract_concatenations(right)]
+    [wrap_concatenation(left, :left) | extract_concatenations(right)]
   end
 
   defp extract_concatenations(other) do
-    [wrap_concatenation(other)]
+    [wrap_concatenation(other, :right)]
   end
 
-  defp wrap_concatenation(binary) when is_binary(binary) do
+  defp wrap_concatenation(binary, _side) when is_binary(binary) do
     binary
   end
 
-  defp wrap_concatenation(other) do
-    {:::, [], [other, {:binary, [], nil}]}
+  defp wrap_concatenation(other, side) do
+    {:::, [],
+     [
+       {:expand_concatenation_argument, [context: Elixir, import: Kernel], [other, side]},
```

Comment:
@ggcampinho theoretically you don't need a new macro. The argument on the expand macro will be exactly the same as you have here.